### PR TITLE
Cached random access to CUs and DIEs

### PR DESCRIPTION
**Pull request by mdmillerii**
Here is a series that refactors the DIE cache to cache entries without walking from the top of a CU,
and also one that adds a similar cache to the CU lookup in the DWARFInfo.

### CODE REVIEW

**Comment by mdmillerii** [comment]
I've been working on this for a few weeks now, developing a test to locate static variables.  I've been testing using a full Linux kernel build so testing a few is useful.

This part has been refactored enough that I'm comfortable pushing it; the bigger script still needs a few rounds.

This addresses #262 so I've put some efffort into getting this commented and pushed.  I just added the LUTEntry object today.

**Comment by eliben** [comment]
Thanks for the contribution. Do you think it would be possible to split this PR into two? One fir DIE cache and one for CU cache?

@sevaa would you like to help review this, since it addresses #262?

**Comment by sevaa** [comment]
`_get_cached_DIE` assumes that the _diemap of a CU already has at least one entry. `get_DIE_from_refaddr` eventually calls `_get_cached_DIE` for the target CU.

I'm dealing with cases of both intra-CU and cross-CU references (Xcode **loves** the latter). If one tries to follow the reference to another CU where not a single DIE has been parsed/cached so far, will that work? Please test the scenario.

**Comment by mdmillerii** [comment]
@sevaa  does a4fb515473e300ebaa5cc76bd04563656dfb5278 address your concern?

I left this as two commits to show the code motion for review but will squash as requested by Eli.

**Comment by sevaa** [comment]
It does.

**Comment by mdmillerii** [comment]
> Do you think it would be possible to split this PR into two? One fir DIE cache and one for CU cache?

I initially read this as can I factor this into two commits, but then I realized you asked for multiple PRs.  
It would take at least 3 to get back to this state:  
1. The first two commits (the get_parent and the DIE by ref within a CU) can be squashed.  
2. CU caching is the third commmit
3. The DIE API helpers and the testcase can be squashed.

Based on your request I squashed DIE cache cleanup (which was always just showing my work but caused review noise), and consolidated the find a DIE pieces into a helpers commit.

**Comment by eliben** [comment]
@rhelmot @rupran @woodruffw pinging some folks who've been active recently -- since this is a large change, I'd really appreciate some help with reviewing :-) My review queue for pycparser has grown rather long.

**Comment by eliben** [comment]
@mdmillerii do you have some measurements / benchmark numbers to report? What kind of operations does this make faster, and how much faster?

Note to self: we really need some portfolio of files for performance measurements

**Comment by woodruffw** [comment]
I can review (and test) here in a bit, thanks for the ping!

**Comment by eliben** [comment]
> I can review (and test) here in a bit, thanks for the ping!

Thanks! I'll assign you, just unassign when you're done

File: elftools/dwarf/dwarfinfo.py [hunk 1]
```diff
@@ -112,6 +117,84 @@ def has_debug_info(self):
         """
         return bool(self.debug_info_sec)
 
+    def get_DIE_from_lute(self, lutentry):
```
Comment (woodruffw) [inline elftools/dwarf/dwarfinfo.py]: Nit: Maybe rename to either `get_DIE_from_lut` or `get_DIE_from_lut_entry`?

**Comment by woodruffw** [comment]
Confirmed that this works with my internal tooling. I'll do some additional performance testing.

**Comment by eliben** [comment]
@woodruffw out of curiosity, what is this big suite of projects you have? Do you have a setup to analyze many different ELF binaries with pyelftools?

File: elftools/dwarf/dwarfinfo.py (hunk 1 above)
Comment (mdmillerii) [reply elftools/dwarf/dwarfinfo.py]: @eliben want me to fix this up and rebase?  (I'd choose lut_entry) .. then I can rebase @265 on top

File: elftools/dwarf/die.py [hunk 2]
```diff
@@ -142,6 +167,40 @@ def set_parent(self, die):
 
     #------ PRIVATE ------#
 
+    def _search_for_ancestors(self):
+        """ Search for our parent by, starting with the CU top die,
+            setting the _parent link of each child and then iterating on
+            the child which was closest to object DIE offset without over,
+            whithan to the object DIE offset, which is an ancestor.
```
Comment (rupran) [inline elftools/dwarf/die.py]: This sentence does not make sense to me, maybe reword or split the sentence in two parts.

File: elftools/dwarf/compileunit.py [hunk 3]
```diff
@@ -98,6 +99,30 @@ def get_top_DIE(self):
...
+            When using a reference class attribute with a form that is
+            relative to the compile unit, add unit add the compile unit's
+            .cu_addr before calling this function.
+        """
+        # All DIEs are after the cu header and within the unit
+        dwarf_assert(
+            self.cu_die_offset <= refaddr < self.cu_offset + self.size,
+            'refaddr %s not in DIE range of CU %s' % (refaddr, self.cu_offset))
+
+        die = self._get_cached_DIE(refaddr)
```
Comment (eliben) [inline elftools/dwarf/compileunit.py]: Maybe just return this directly without the `die =` intermediate?

File: elftools/dwarf/die.py [hunk 4]
```diff
@@ -142,6 +167,40 @@ def set_parent(self, die):
...
+        """
+        search = self.cu.get_top_DIE()
+
+        while search.offset < self.offset:
+
```
Comment (eliben) [inline elftools/dwarf/die.py]: Nit: remove this empty line

File: elftools/dwarf/die.py [hunk 5]
```diff
@@ -142,6 +167,40 @@ def set_parent(self, die):
...
+            # If we didn't find a closer parent, give up, don't loop.
+            # Either we mis-parsed an ancestor or someone asked for a DIE
+            # by an offset that was not actually the start of a DIE.
+            if prev is search:
+                raise ValueError("offset %s not in CU %s DIE tree" %
+                    (self.offset, self.cu.cu_offset))
+
+            search = prev
+
+        pass
```
Comment (eliben) [inline elftools/dwarf/die.py]: Not sure what's going on here. `pass` outside the loop? 

Should this line be removed?

Also the commented code following it

File: elftools/dwarf/dwarfinfo.py (hunk 1 above)
Comment (eliben) [reply elftools/dwarf/dwarfinfo.py]: lut_entry SGTM

File: elftools/dwarf/dwarfinfo.py [hunk 6]
```diff
@@ -253,23 +336,41 @@ def range_lists(self):
 
     #------ PRIVATE ------#
 
-    def _parse_CUs_iter(self):
+    def _parse_CUs_iter(self, offset=0):
```
Comment (eliben) [inline elftools/dwarf/dwarfinfo.py]: Document what offset means

File: elftools/dwarf/dwarfinfo.py [hunk 7]
```diff
@@ -253,23 +336,41 @@ def range_lists(self):
...
+
             yield cu
 
+    def _cached_CU_at_offset(self, offset):
+        """ Lookup cu in the cache.  If not present, first parse and insert it.
+
+            See get_CU_at().
+        """
+        # If the insert point is 0 the entry does not match (perhaps the cache
+        # is empty).  Otherwise on a miss parse and insert.
```
Comment (eliben) [inline elftools/dwarf/dwarfinfo.py]: I'm not sure I understand what the second sentence in this comment means. Please rephrase.

File: test/test_dwarf_cu_and_die_cache.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+                               'lambda.elf'), 'rb') as f:
+            elffile = ELFFile(f)
+            self.assertTrue(elffile.has_dwarf_info())
+
+            dwarf = elffile.get_dwarf_info()
+            pt = dwarf.get_pubnames()
+            for (k, v) in pt.items():
+                ndie = dwarf.get_DIE_from_lute(v)
+                self.dprint(ndie)
+                # if not ndie.tag == 'DW_TAG_variable':
```
Comment (eliben) [inline test/test_dwarf_cu_and_die_cache.py]: Remove?

**Comment by woodruffw** [comment]
> out of curiosity, what is this big suite of projects you have? Do you have a setup to analyze many different ELF binaries with pyelftools?

Sorry for the ambiguity there: it's both a big suite of projects (i.e. test binaries) _and_ tools that use pyelftools to extract information from those binaries.

There's nothing publicly available at the moment, but it's all part of the [DARPA CHESS](https://www.darpa.mil/program/computers-and-humans-exploring-software-security) program.

File: elftools/dwarf/die.py (hunk 2 above)
Comment (mdmillerii) [reply elftools/dwarf/die.py]: yes, that was trying to stuff everything into a sentence and it got lost.

File: elftools/dwarf/compileunit.py (hunk 3 above)
Comment (mdmillerii) [reply elftools/dwarf/compileunit.py]: done

File: elftools/dwarf/die.py (hunk 4 above)
Comment (mdmillerii) [reply elftools/dwarf/die.py]: done

File: elftools/dwarf/die.py (hunk 5 above)
Comment (mdmillerii) [reply elftools/dwarf/die.py]: refactoring remenants, fixed.

File: elftools/dwarf/dwarfinfo.py (hunk 7 above)
Comment (mdmillerii) [reply elftools/dwarf/dwarfinfo.py]: it was saying if the first sentence was false, then do the actions.  Hopefully the new is clearer,

**Comment by mdmillerii** [comment]
Here's my update.   I also pushed a branch [random-die-review](http://github.com/mdmillerii/pyelftools/tree/random-die-review) that shows the previous content rebased then 1 patch reverted so reviewers can see what was changed, specifically on https://github.com/eliben/pyelftools/commit/a5d42a4ea9fa144ba027ecaee9f8778b177ff6fa
